[Daniel Tan] Duke Increments

CarbonGrid opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

CarbonGrid committed:
Constructed initial skeleton of Duke

CarbonGrid committed:
Added ability to store user input and re-display when requested.

CarbonGrid committed:
Added ability to mark tasks as done & Improved Handling of output.

CarbonGrid committed:
Added ToDos, Events, Deadlines & Improved Input Handling

CarbonGrid committed:
Added text UI testing using i/o redirection

CarbonGrid committed:
Added Handling of errors with DukeException

CarbonGrid committed:
Improved handling for incorrect inputs.

CarbonGrid committed:
Added support for deleting tasks from list.

CarbonGrid committed:
+ BUGFIX: Unhandled exception when "done 1" on empty
+ Classified Generic Exceptions

CarbonGrid committed:
Implemented Level-7 save feature

CarbonGrid committed:
Implemented Level-8 featuring Dates recognition

CarbonGrid committed:
Merge branch 'branch-level-7'

CarbonGrid committed:
+ Merge branch 'branch-level-8'
+ Resolved conflict with master
+ Done proper datetime formatting in file required by Level-8
+ Fixed FileIO bug that happens when list goes empty

CarbonGrid committed:
+ Refactored code to model OOP more closely
+ Corrected time default of datetime parsing

CarbonGrid committed:
Merge branch 'gradle'

CarbonGrid committed:
+ Added Gradle Support
+ Added JUnit tests
+ Corrected Wrongly Named Exception

CarbonGrid committed:
+ Added JavaDoc

CarbonGrid committed:
+ Adopted coding standard as per defined by gradle plugin checkstyle v8.23

CarbonGrid committed:
+ Implemented Level-9: Find from TaskList

CarbonGrid committed:
Merge branch 'branch-A-JavaDoc'

CarbonGrid committed:
Merge branch 'branch-A-CodingStandard'
+ Resolved Merge conflicts

CarbonGrid committed:
Merge branch 'branch-Level-9'
+ Resolved Merge Conflicts

CarbonGrid committed:
+ Organized the classes into suitable java packages.

CarbonGrid committed:
+ Fixed non-compliance to coding standard
+ Fixed vowel detection in EmptyFieldDukeException

shihaoyap commented:
Hi Daniel! While looking through your program, i did not find any coding style violations. Well done! :) I found your code structure to be really neat and readable and it follows good OOP principles. I also like the fact that you did prepare for multiple scenarios of invalid input function, and your program was able to handle the different input exceptions really well, when i was testing it 👍 

Overall the program was well written! I tried hard to break it but was not able to haha! Well done! i look forward to seeing your program improve! 

Suggestions - Could explore creating a UI (as stated in Level-10) to be used to interact with the users.

CarbonGrid committed:
+ Implemented Level-10: Duke's GUI

CarbonGrid committed:
Merge branch 'branch-Level-10'

CarbonGrid commented:
> Hi Daniel! While looking through your program, i did not find any coding style violations. Well done! :) I found your code structure to be really neat and readable and it follows good OOP principles. I also like the fact that you did prepare for multiple scenarios of invalid input function, and your program was able to handle the different input exceptions really well, when i was testing it 👍
> 
> Overall the program was well written! I tried hard to break it but was not able to haha! Well done! i look forward to seeing your program improve!
> 
> Suggestions - Could explore creating a UI (as stated in Level-10) to be used to interact with the users.

@shihaoyap Thanks for the in-depth review. 

Happy to meet the expectations. :smiley: 
Sorry for the letdown. 
Couldn't commit Level-10 early cause still :mask:

CarbonGrid committed:
+ Added Welcome Message
+ Fixed: Duke Ui should terminate at "bye"

CarbonGrid committed:
Merge branch 'branch-Level-10'

CarbonGrid committed:
+ Fixed non-compliance to coding standard from Level-10 merge.

CarbonGrid committed:
+ Implemented Assertions and Improved code quality.

CarbonGrid committed:
Merge pull request #4 from CarbonGrid/branch-A-Assertions

+ Implemented Assertions and Improved code quality.

CarbonGrid committed:
Travis build test

CarbonGrid committed:
Permission access for travis

CarbonGrid committed:
Update README.md

CarbonGrid committed:
+ Implement D-Places with available command "place" & "alias". (#5)

Using squash & merge.

CarbonGrid committed:
+ Adopted Serialization using Standard Libraries instead.
+ Places are now stored/saved.

lrchema commented on src/main/java/Launcher.java:
```diff
@@ -0,0 +1,8 @@
+import javafx.application.Application;
+
```
please consider adding javadoc comments to this class

CarbonGrid replied:
I was considering to remove this class. After all, it's a Launcher that does nothing.

sianghwee commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    @Override
+    public void start(Stage stage) {
+        assert stage != null;
```
Give a description/comment for your assert to aid troubleshooting

CarbonGrid replied:
I guess it differs from person to person. But IMO, it should be self-explanatory enough that I wouldn't think of placing any comment.

sianghwee replied:
Hello, I think I was not clear in my initial comment. What I meant was that when you assert, perhaps you can do `assert stage != null : "Stage not found"; ` as that will make it clear what the problem is when your assert fails.

CarbonGrid replied:
😄 I was wondering if you meant an assert message, which is great. But I concluded you suggest me to leave a comment by the assert.

sianghwee commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,76 @@
+import duke.DialogBox;
+import duke.Duke;
+
+import duke.MainUi;
```
It's good that you never use wildcard imports but it would be better if you grouped the imports by packages.

CarbonGrid replied:
Not sure when that line break got in. But good observation.

sianghwee commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,76 @@
+            .add(DialogBox.getDukeDialog(new Label(tmpUi.getResponse()), new ImageView(dukeImage)));
+    }
+
+    public void setDuke(Duke d) {
```
Write a javadoc to explain this method.

CarbonGrid replied:
IMO, this is self-explanatory and simple. Sometimes it's hard for me to justify whether getters & setters deserve JavaDoc.

sianghwee commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,90 @@
+     * @param l  String textContent
+     * @param iv Profile Image for DialogBox
+     */
+    public DialogBox(Label l, ImageView iv) {
```
I feel that it might be better if you use variable names that are more descriptive especially for such a long method.

CarbonGrid replied:
Justified.

sianghwee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,80 @@
+        } catch (DukeException e) {
+            ui.show(e.getMessage());
+        } finally {
+            return ((MainUi) ui).getResponse();
```
Why did you need to typecast your Ui?

CarbonGrid replied:
MainWindow uses MainUi while Ui doesn't have any definition of getResponse(). There is alternative ways of cleaner implementation but I believe but I thought it would be simpler to cast it. Might put this on forum for discussion on good practice for non-existent subclass methods in super-class.

sianghwee commented on src/main/java/duke/MainUi.java:
```diff
@@ -0,0 +1,21 @@
+package duke;
```
Perhaps you can consider having a separate sub-package (duke.ui) for all your Ui to improve readability.

CarbonGrid replied:
Observant enough. 😄  I did the rest of the package structure in such a way that inheriting class under parent class is of same sub-package. However, forgotten about MainUi extends Ui.

lrchema commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,18 @@
+import duke.Ui;
+import duke.exception.DukeException;
+
+public class FindCommand extends Command {
```
I think it would be good to add javadoc comments here

CarbonGrid replied:
I'm not sure if it is needed here. But I think I should at least have a Javadoc at least in the constructor or execute().
After all one would know what FindCommand probably is from inheritance but one wouldn't know what it does on execution.

blimyj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,60 @@
+            Place.setRefLatLong((HashMap) ois.readObject());
+            Place.setRefAlias((HashMap) ois.readObject());
+            return (TaskList) ois.readObject();
+        } catch (Exception e) {
```
I don't think it's a good idea to catch a general exception. For example, if it is a FileNotFoundException, a new file could be created with a clearer error message printed.

CarbonGrid replied:
Agreed. Scope should always be minimal.

sianghwee commented on src/main/java/duke/MainUi.java:
```diff
@@ -0,0 +1,21 @@
+package duke;
+
+public class MainUi extends Ui {
```
Include a Javadoc for your classes as it is unclear what each class does.

CarbonGrid replied:
Without constructor JavaDoc, I believe that is necessary. Good suggestion.

lrchema commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        assert str != null;
+
+        str = str.trim();
+        int lastIndex = str.indexOf(' ');
```
This seems a little complicated. Can you achieve the same outcome by using .split()?

CarbonGrid replied:
Definitely yes, I did a bit of hand-optimization here. As you can see, split is only good for one word. As you know split would process split at all white-spaces. But I just figured I could implement using lazy look-ahead. Hence, great thought from you that I would improve.

sianghwee commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,88 @@
+        case "list":
+            return new ListCommand();
+        case "deadline":
+            temp = str.substring(8)
```
Perhaps you can try not to use magic literal in your code as it will be more intuitive for the next developer.

CarbonGrid replied:
Bad practice indeed.

sianghwee commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @throws LoadFileFailDukeException if file or raw data can't be loaded
+     */
+    @SuppressWarnings("unchecked")
```
Perhaps you can try to resolve the warning instead of suppressing the warning as it might lead to more underlying bugs.  Is there a reason for this decision? 😄

CarbonGrid replied:
I tried searching for an appropriate way to handle reading Objects. Unfortunately, all solutions I have come across so far has no "clean" way of dealing with Object casting as there is an implied order of processing.

sianghwee reviewed (commented):
Sorry, we were only able to review to the Storage.java as we were pressed for time. Thank you for sharing your code and we were able to learn a lot from it.

CarbonGrid committed:
+ Improved code as per peer review.

CarbonGrid committed:
+ Constructs User Guide
+ BUGFIX: DateTimeFormatter

Set theme jekyll-theme-midnight

Set theme jekyll-theme-hacker

Set theme jekyll-theme-midnight

FIX: Broken Build found by Travis

Improves User Guide

CarbonGrid committed:
Updates build.gradle

CarbonGrid committed:
Updates User Guide
